plugins/fileservices/fileservices.cpp

```diff
@@ -198,63 +198,123 @@ static IPropertyTree *getEnvironmentTree(IConstEnvironment * daliEnv)
+                    {
                         // if any enabled feature has service ws_fs then use this binding
                         if (iter3->query().getPropBool("@authenticate")&&
-                            iter3->query().getProp("@service",tmp.clear())&&
-                            (strcmp(tmp.str(),"ws_fs")==0)) {
-                            if (iter2->query().getProp("@protocol",tmp.clear())) {
-                                tmp.append("://");
+                            iter3->query().getProp("@service",wsfsurl.clear())&&
+                            (strcmp(wsfsurl.str(),"ws_fs")==0)){
+                            if (iter2->query().getProp("@protocol",wsfsurl.clear())) {
+                                wsfsurl.append("://");
                                 StringBuffer espname;
                                 if (iter1->query().getProp("@name",espname)) {
                                     StringBuffer espinst;
                                     if (iter1->query().getProp("Instance[1]/@computer",espinst)) {
                                         StringBuffer ipq;
-                                        if (env->getProp(ipq.appendf("Hardware/Computer[@name=\"%s\"]/@netAddress",espinst.str()).str(),tmp)) {
-                                            tmp.append(':').append(iter2->query().getPropInt("@port",8010)).append("/FileSpray"); // FileSpray seems to be fixed
-                                            espurl.set(tmp);
-                                            PROGLOG("fileservices using esp URL: %s",espurl.get());
-                                            break;
+                                        if (env->getProp(ipq.appendf("Hardware/Computer[@name=\"%s\"]/@netAddress",espinst.str()).str(),wsfsurl))
+                                        {
+                                            wsfsurl.append(':').append(iter2->query().getPropInt("@port",8010)).append("/FileSpray"); // FileSpray seems to be fixed
+                                            addConfiguredWsFSUrl(wsfsurl.str());
+                                            break; //There can be multiple AuthFeatures per ESP Instance
                                         }
                                     }
                                 }
                             }
                         }
-                    }
-                    if (!espurl.isEmpty()) 
-                        break;
-                }
-                if (!espurl.isEmpty()) 
-                    break;
-            }
+                    }//AuthenticateFeature
+                }//ESPBinding
+            }//ESPProcess
         }
+        if (isUrlListEmpty())
+            throw MakeStringException(-1,"Could not find any WS FileSpray in the target HPCC configuration.");
     }
-    if (espurl.isEmpty()) 
-        throw MakeStringException(-1,"Cannot determine ESP Url");
-    return espurl.get();
-}
 
+    const char * wsfsurl = getNextAliveWsFSURL(wu);
```

Comment:
Bit of a shame to start from beginning of list each time rather than last successful connection (though saving last successful connection per user might be a bit of a pain).